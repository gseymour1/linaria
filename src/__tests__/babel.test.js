/* eslint-disable no-template-curly-in-string */
/* @flow */

const babel = require('@babel/core');
const dedent = require('dedent');

const transpile = async input => {
  const { code } = await babel.transformAsync(input, {
    presets: [[require.resolve('../babel'), { evaluate: false }]],
    plugins: ['@babel/plugin-syntax-jsx'],
    filename: '/app/index.js',
  });

  return code;
};

it('transpiles styled template literal with object', async () => {
  const code = await transpile(
    dedent`
    const Title = styled.h1\`
      font-size: 14px;
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});

it('transpiles styled template literal with function and tag', async () => {
  const code = await transpile(
    dedent`
    const Title = styled('h1')\`
      font-size: 14px;
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});

it('transpiles styled template literal with function and component', async () => {
  const code = await transpile(
    dedent`
    const Title = styled(Heading)\`
      font-size: 14px;
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});

it('outputs valid CSS classname', async () => {
  const code = await transpile(
    dedent`
    const á¾©Page$Title = styled.h1\`
      font-size: 14px;
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});

it('evaluates and inlines expressions in scope', async () => {
  const code = await transpile(
    dedent`
    const color = 'blue';

    const Title = styled.h1\`
      color: ${'${color}'};
      width: ${'${100 / 3}'}%;
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});

it('inlines object styles as CSS string', async () => {
  const code = await transpile(
    dedent`
    const cover = {
      position: 'absolute',
      top: 0,
      right: 0,
      bottom: 0,
      left: 0,
      opacity: 1,
      minHeight: 420,

      '@media (min-width: 200px)': {
        WebkitOpacity: .8,
        MozOpacity: .8,
        msOpacity: .8,
        OOpacity: .8,
        WebkitBorderRadius: 2,
        MozBorderRadius: 2,
        msBorderRadius: 2,
        OBorderRadius: 2,
        WebkitTransition: '400ms',
        MozTransition: '400ms',
        OTransition: '400ms',
        msTransition: '400ms',
      }
    };

    const Title = styled.h1\`
      ${'${cover}'}
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});

it('replaces unknown expressions with CSS custom properties', async () => {
  const code = await transpile(
    dedent`
    const Title = styled.h1\`
      font-size: ${'${size}'}px;
      color: ${'${props => props.color}'};
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});

it('handles interpolation followed by unit', async () => {
  const code = await transpile(
    dedent`
    const Title = styled.h1\`
      font-size: ${'${size}'}em;
      text-shadow: black 1px ${'${shadow}'}px, white -2px -2px;
      margin: ${'${size}'}px;
      width: calc(2 * ${'${props => props.width}'}vw);
      height: ${'${props => { if (true) { return props.height } else { return 200 } }}'}px;
      grid-template-columns: ${'${unit}'}fr 1fr 1fr ${'${unit}'}fr;
      border-radius: ${'${function(props) { return 200 }}'}px
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});

it('uses the same custom property for the same identifier', async () => {
  const code = await transpile(
    dedent`
    const Box = styled.div\`
      height: ${'${size}'}px;
      width: ${'${size}'}px;
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});

it('uses the same custom property for the same expression', async () => {
  const code = await transpile(
    dedent`
    const Box = styled.div\`
      height: ${'${props => props.size}'}px;
      width: ${'${props => props.size}'}px;
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});

it('handles nested blocks', async () => {
  const code = await transpile(
    dedent`
    const Button = styled.button\`
      font-family: ${'${regular}'};

      &:hover {
        border-color: blue;
      }

      @media (max-width: 200px) {
        width: 100%;
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});

it('prevents class name collision', async () => {
  const code = await transpile(
    dedent`
    const Title = styled.h1\`
      font-size: ${'${size}'}px;
      color: ${'${props => props.color}'}
    \`;

    function more() {
      const Title = styled.h1\`
        font-family: ${'${regular}'};
      \`;
    }
    `
  );

  expect(code).toMatchSnapshot();
});

it('throws when not attached to a variable', async () => {
  expect.assertions(1);

  try {
    await transpile(
      dedent`
      styled.h1\`
        font-size: ${'${size}'}px;
        color: ${'${props => props.color}'}
      \`;
      `
    );
  } catch (e) {
    expect(e.message).toMatchSnapshot();
  }
});

it('does not output CSS if none present', async () => {
  const code = await transpile(
    dedent`
      const number = 42;

      const title = String.raw\`This is something\`;
      `
  );

  expect(code).toMatchSnapshot();
});

it('transpiles css template literal', async () => {
  const code = await transpile(
    dedent`
    const title = css\`
      font-size: 14px;
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});

it('handles css template literal in object property', async () => {
  const code = await transpile(
    dedent`
    const components = {
      title: css\`
        font-size: 14px;
      \`
    };
    `
  );

  expect(code).toMatchSnapshot();
});

it('handles css template literal in JSX element', async () => {
  const code = await transpile(
    dedent`
    <Title class={css\` font-size: 14px; \`} />
    `
  );

  expect(code).toMatchSnapshot();
});

it('throws when contains dynamic expression without evaluate: true in css tag', async () => {
  expect.assertions(1);

  try {
    await transpile(
      dedent`
      const title = css\`
        font-size: ${'${size}'}px;
      \`;
      `
    );
  } catch (e) {
    expect(e.message).toMatchSnapshot();
  }
});
