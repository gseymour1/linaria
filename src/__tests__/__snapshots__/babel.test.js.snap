// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not output CSS if none present 1`] = `
"const number = 42;
const title = String.raw\`This is something\`;"
`;

exports[`evaluates and inlines expressions in scope 1`] = `
"const color = 'blue';
const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_th73eyu\\"
});

/*
CSS OUTPUT TEXT START
.Title_th73eyu{color:blue;width:33.333333333333336%;}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":3,\\"column\\":6},\\"name\\":\\"Title_th73eyu\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;

exports[`handles css template literal in JSX element 1`] = `
"<Title class={\\"Title_th73eyu\\"} />;

/*
CSS OUTPUT TEXT START
.Title_th73eyu{font-size:14px;}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":1,\\"column\\":13},\\"name\\":\\"Title_th73eyu\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;

exports[`handles css template literal in object property 1`] = `
"const components = {
  title: \\"title_th73eyu\\"
};

/*
CSS OUTPUT TEXT START
.title_th73eyu{font-size:14px;}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":2,\\"column\\":2},\\"name\\":\\"title_th73eyu\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;

exports[`handles interpolation followed by unit 1`] = `
"const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_th73eyu\\",
  vars: {
    \\"th73eyu-0\\": [size, \\"em\\"],
    \\"th73eyu-1\\": [shadow, \\"px\\"],
    \\"th73eyu-2\\": [size, \\"px\\"],
    \\"th73eyu-3\\": [props => props.width, \\"vw\\"],
    \\"th73eyu-4\\": [props => {
      if (true) {
        return props.height;
      } else {
        return 200;
      }
    }, \\"px\\"],
    \\"th73eyu-5\\": [unit, \\"fr\\"],
    \\"th73eyu-7\\": [function (props) {
      return 200;
    }, \\"px\\"]
  }
});

/*
CSS OUTPUT TEXT START
.Title_th73eyu{font-size:var(--th73eyu-0);text-shadow:black 1px var(--th73eyu-1),white -2px -2px;margin:var(--th73eyu-2);width:calc(2 * var(--th73eyu-3));height:var(--th73eyu-4);grid-template-columns:var(--th73eyu-5) 1fr 1fr var(--th73eyu-5);border-radius:var(--th73eyu-7);}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":1,\\"column\\":6},\\"name\\":\\"Title_th73eyu\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;

exports[`handles nested blocks 1`] = `
"const Button =
/*#__PURE__*/
styled(\\"button\\")({
  name: \\"Button\\",
  class: \\"Button_bh73eyu\\",
  vars: {
    \\"bh73eyu-0\\": [regular]
  }
});

/*
CSS OUTPUT TEXT START
.Button_bh73eyu{font-family:var(--bh73eyu-0);}.Button_bh73eyu:hover{border-color:blue;}@media (max-width:200px){.Button_bh73eyu{width:100%;}}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":1,\\"column\\":6},\\"name\\":\\"Button_bh73eyu\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;

exports[`inlines object styles as CSS string 1`] = `
"const cover = {
  position: 'absolute',
  top: 0,
  right: 0,
  bottom: 0,
  left: 0,
  opacity: 1,
  minHeight: 420,
  '@media (min-width: 200px)': {
    WebkitTransition: '400ms',
    MozTransition: '400ms',
    OTransition: '400ms',
    msTransition: '400ms'
  }
};
const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_th73eyu\\"
});

/*
CSS OUTPUT TEXT START
.Title_th73eyu{position:absolute;top:0;right:0;bottom:0;left:0;opacity:1;min-height:420px;}@media (min-width:200px){.Title_th73eyu{-webkit-transition:400ms;-moz-transition:400ms;-o-transition:400ms;-ms-transition:400ms;}}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":18,\\"column\\":6},\\"name\\":\\"Title_th73eyu\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;

exports[`prevents class name collision 1`] = `
"const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_th73eyu\\",
  vars: {
    \\"th73eyu-0\\": [size, \\"px\\"],
    \\"th73eyu-1\\": [props => props.color]
  }
});

function more() {
  const Title =
  /*#__PURE__*/
  styled(\\"h1\\")({
    name: \\"Title\\",
    class: \\"Title_t462dh3\\",
    vars: {
      \\"t462dh3-0\\": [regular]
    }
  });
}

/*
CSS OUTPUT TEXT START
.Title_th73eyu{font-size:var(--th73eyu-0);color:var(--th73eyu-1);}
.Title_t462dh3{font-family:var(--t462dh3-0);}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":1,\\"column\\":6},\\"name\\":\\"Title_th73eyu\\"},{\\"generated\\":{\\"line\\":2,\\"column\\":0},\\"original\\":{\\"line\\":7,\\"column\\":8},\\"name\\":\\"Title_t462dh3\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;

exports[`replaces unknown expressions with CSS custom properties 1`] = `
"const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_th73eyu\\",
  vars: {
    \\"th73eyu-0\\": [size, \\"px\\"],
    \\"th73eyu-1\\": [props => props.color]
  }
});

/*
CSS OUTPUT TEXT START
.Title_th73eyu{font-size:var(--th73eyu-0);color:var(--th73eyu-1);}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":1,\\"column\\":6},\\"name\\":\\"Title_th73eyu\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;

exports[`throws when contains dynamic expression without evaluate: true in css tag 1`] = `
"/app/index.js: The CSS cannot contain JavaScript expressions when using the 'css' tag. To evaluate the expressions at build time, pass 'evaluate: true' to the babel plugin.
[0m [90m 1 | [39m[36mconst[39m title [33m=[39m css[32m\`[39m[0m
[0m[31m[1m>[22m[39m[90m 2 | [39m[32m  font-size: \${size}px;[39m[0m
[0m [90m   | [39m               [31m[1m^[22m[39m[0m
[0m [90m 3 | [39m[32m\`[39m[33m;[39m[0m"
`;

exports[`throws when not attached to a variable 1`] = `
"/app/index.js: Couldn't determine a name for the component. Ensure that it's either:
- Assigned to a variable
- Is an object property
- Is a prop in a JSX element

[0m[31m[1m>[22m[39m[90m 1 | [39mstyled[33m.[39mh1[32m\`[39m[0m
[0m [90m   | [39m[31m[1m^[22m[39m[0m
[0m [90m 2 | [39m[32m  font-size: \${size}px;[39m[0m
[0m [90m 3 | [39m[32m  color: \${props => props.color}[39m[0m
[0m [90m 4 | [39m[32m\`[39m[33m;[39m[0m"
`;

exports[`transpiles css template literal 1`] = `
"const title = \\"title_th73eyu\\";

/*
CSS OUTPUT TEXT START
.title_th73eyu{font-size:14px;}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":1,\\"column\\":6},\\"name\\":\\"title_th73eyu\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;

exports[`transpiles styled template literal with function and component 1`] = `
"const Title =
/*#__PURE__*/
styled(Heading)({
  name: \\"Title\\",
  class: \\"Title_th73eyu\\"
});

/*
CSS OUTPUT TEXT START
.Title_th73eyu{font-size:14px;}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":1,\\"column\\":6},\\"name\\":\\"Title_th73eyu\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;

exports[`transpiles styled template literal with function and tag 1`] = `
"const Title =
/*#__PURE__*/
styled('h1')({
  name: \\"Title\\",
  class: \\"Title_th73eyu\\"
});

/*
CSS OUTPUT TEXT START
.Title_th73eyu{font-size:14px;}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":1,\\"column\\":6},\\"name\\":\\"Title_th73eyu\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;

exports[`transpiles styled template literal with object 1`] = `
"const Title =
/*#__PURE__*/
styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"Title_th73eyu\\"
});

/*
CSS OUTPUT TEXT START
.Title_th73eyu{font-size:14px;}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":1,\\"column\\":6},\\"name\\":\\"Title_th73eyu\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;

exports[`uses the same custom property for the same expression 1`] = `
"const Box =
/*#__PURE__*/
styled(\\"div\\")({
  name: \\"Box\\",
  class: \\"Box_bh73eyu\\",
  vars: {
    \\"bh73eyu-0\\": [props => props.size, \\"px\\"]
  }
});

/*
CSS OUTPUT TEXT START
.Box_bh73eyu{height:var(--bh73eyu-0);width:var(--bh73eyu-0);}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":1,\\"column\\":6},\\"name\\":\\"Box_bh73eyu\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;

exports[`uses the same custom property for the same identifier 1`] = `
"const Box =
/*#__PURE__*/
styled(\\"div\\")({
  name: \\"Box\\",
  class: \\"Box_bh73eyu\\",
  vars: {
    \\"bh73eyu-0\\": [size, \\"px\\"]
  }
});

/*
CSS OUTPUT TEXT START
.Box_bh73eyu{height:var(--bh73eyu-0);width:var(--bh73eyu-0);}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":1,\\"column\\":6},\\"name\\":\\"Box_bh73eyu\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;
